plugins/modules/vmware_deploy_ovf.py

```diff
@@ -217,10 +241,95 @@ def path_exists(value):
         value = str(value)
     value = os.path.expanduser(os.path.expandvars(value))
     if not os.path.exists(value):
-        raise ValueError('%s is not a valid path' % value)
+        raise ValueError(f"'{value}' is not a valid path")
     return value
 
 
+class WebHandle(object):
+    def __init__(self, url):
+        self.url = url
+        self.thumbprint = None
+        self.ssl_context = None
+
+        self.parsed_url = self._parse_url(url)
+
+        self.https = self.parsed_url.group('scheme') == 'https://'
+
+        if self.https:
+            self.ssl_context = ssl._create_default_https_context()
+            self.ssl_context.check_hostname = False
+            self.ssl_context.verify_mode = ssl.CERT_NONE
+
+            self.thumbprint = self._get_thumbprint(
+                self.parsed_url.group('hostname'))
+            r = urlopen(url=url, context=self.ssl_context)
+        else:
+            r = urlopen(url)
+        if r.code != 200:
+            raise FileNotFoundError(url)
+        self.headers = self._headers_to_dict(r)
+        if 'accept-ranges' not in self.headers:
+            raise Exception("Site does not accept ranges")
+        self.st_size = int(self.headers['content-length'])
+        self.offset = 0
+
+    def _parse_url(self, url):
+        exp1 = r"(?P<url>(?:(?P<scheme>[a-zA-Z]+:\/\/)?(?P<hostname>(?:[-a-zA-Z0-9@%_\+~#=]{1,256}\.){1,256}(?:[-a-zA-Z0-9@%_\+~#=]{1,256})))"
+        exp2 = r"(?::(?P<port>[[:digit:]]+))?(?P<path>(?:\/[-a-zA-Z0-9!$&'()*+,\\\/:;=@\[\]._~%]*)*)(?P<query>(?:(?:\#|\?)[-a-zA-Z0-9!$&'()*+,\\\/:;=@\[\]._~]*)*))"
```

Comment:
I hope I'll never have to troubleshoot this :laughing:

Reply:
It's actually a pretty simple regular expression. =) It beats any URL to the full set of its constituent parts (scheme, hostname, path, etc)

Reply:
Well, I think I'm not that bad at regular expressions... but if you think this simple, I don't want to see one that you consider complex :stuck_out_tongue: SCNR

Reply:
I'm afraid we just need to catch our luck with CI again =(